support-frontend/app/controllers/RedemptionController.scala

```diff
@@ -220,19 +237,40 @@ class TestUserFromRequest(identityService: IdentityService, testUsers: TestUserS
 
 }
 
-class GetCorporateCustomer(dynamoLookup: DynamoTableAsyncForUser) {
+class CodeValidator(zuoraLookupService: ZuoraGiftLookupService, dynamoLookup: DynamoTableAsyncForUser) {
+
+  def validate(inputCode: String, isTestUser: Boolean)(implicit ec: ExecutionContext): EitherT[Future, String, ReaderType] =
+    EitherT(getValidationResult(inputCode, isTestUser).map {
+      case ValidCorporateCode(_) => Right(Corporate)
+      case ValidGiftCode(_) => Right(Gift)
+      case invalidCode => Left(validationResultToErrorMessage(invalidCode))
+    })
 
-  def apply(redemptionCode: String, isTestUser: Boolean)(implicit ec: ExecutionContext): EitherT[Future, String, Unit] = {
-    val getCodeStatus = GetCodeStatus.withDynamoLookup(dynamoLookup(isTestUser))
+  private def getValidationResult(inputCode: String, isTestUser: Boolean)(implicit ec: ExecutionContext): Future[CodeValidationResult] = {
+    val corporateValidator = CorporateCodeValidator.withDynamoLookup(dynamoLookup(isTestUser))
+    val giftValidator = new GiftCodeValidator(zuoraLookupService)
 
-    for {
-      codeToCheck <- EitherT.fromEither[Future](RedemptionCode(redemptionCode)).leftMap(_ => "Please check the code and try again")
-      _ <- EitherT(getCodeStatus(codeToCheck)).leftMap {
-        case GetCodeStatus.NoSuchCode => "Please check the code and try again"
-        case GetCodeStatus.CodeAlreadyUsed => "Your code has already been redeemed"
-      }
-    } yield ()
+    RedemptionCode(inputCode)
+      .leftMap(_ => CodeMalformed)
+      .map {
+        redemptionCode =>
+          for {
+            giftValidationResult <- giftValidator.validate(redemptionCode, "")
+            mergedResult <- if (giftValidationResult == CodeNotFound)
+              corporateValidator.validate(redemptionCode)
+            else
+              Future.successful(giftValidationResult)
+          } yield mergedResult
+      }.fold(
+      malformedCode => Future.successful(malformedCode),
+      validationResult => validationResult
+    )
```

Comment:
may be slightly clearer with pattern match?
```suggestion
      } match {
        case Left(malformedCode) => Future.successful(malformedCode),
        case Right(validationResult) => validationResult
    }
```

Reply:
in fact just replace the fold with `.merge`, and replace line 254 with 
```scala
.leftMap(_ => Future.successful(CodeMalformed))
```